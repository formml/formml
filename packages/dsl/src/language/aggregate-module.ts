import { EmptyFileSystem, inject } from 'langium'
import {
  DefaultSharedModuleContext,
  LangiumSharedServices,
  createDefaultModule,
  createDefaultSharedModule,
} from 'langium/lsp'

import { FormMLModule, FormMLServices } from './formml/formml-module.js'
import { registerFormMLValidationChecks } from './formml/formml-validator.js'
import {
  FormMLDeclarationModule,
  FormMLDeclarationServices,
} from './formml-declaration/formml-declaration-module.js'
import {
  FormMLDeclarationGeneratedModule,
  FormMLGeneratedModule,
  FormMLGeneratedSharedModule,
} from './generated/module.js'
import LongestMatchServiceRegistry from './overrides/LongestMatchServiceRegistry.js'

/**
 * Create the full set of services required by Langium.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection
 * @returns An object wrapping the shared services and the language-specific services
 */
export function createAggregateServices(context: DefaultSharedModuleContext): {
  FormML: FormMLServices
  FormMLDeclaration: FormMLDeclarationServices
  shared: LangiumSharedServices
} {
  const shared = inject(
    createDefaultSharedModule(context),
    FormMLGeneratedSharedModule,
    {
      ServiceRegistry: () => new LongestMatchServiceRegistry(),
    },
  )
  const FormML = inject(
    createDefaultModule({ shared }),
    FormMLGeneratedModule,
    FormMLModule,
  )
  const FormMLDeclaration = inject(
    createDefaultModule({ shared }),
    FormMLDeclarationGeneratedModule,
    FormMLDeclarationModule,
  )
  shared.ServiceRegistry.register(FormML)
  shared.ServiceRegistry.register(FormMLDeclaration)

  registerFormMLValidationChecks(FormML)
  // registerFormMLDeclarationValidationChecks(FormMLDeclaration)

  if (!context.connection) {
    // We don't run inside a language server
    // Therefore, initialize the configuration provider instantly
    void shared.workspace.ConfigurationProvider.initialized({})
  }
  return { FormML, FormMLDeclaration, shared }
}

export const createInMemoryAggregateServices = () =>
  createAggregateServices(EmptyFileSystem)
