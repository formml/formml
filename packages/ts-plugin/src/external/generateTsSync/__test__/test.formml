form testForm {
  num  numField
  text textField
}
