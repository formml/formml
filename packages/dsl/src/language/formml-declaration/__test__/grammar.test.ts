import { clearDocuments, parseHelper } from 'langium/test'

import { createInMemoryAggregateServices } from '../../aggregate-module.js'
import { FormMLDeclaration } from '../../generated/ast.js'

describe('formml declaration grammar', () => {
  const services = createInMemoryAggregateServices()

  afterEach(async () => {
    await clearDocuments(services.FormMLDeclaration)
  })

  const serialize = (ast: FormMLDeclaration) =>
    services.FormMLDeclaration.serializer.JsonSerializer.serialize(ast, {
      space: 2,
    })
  const parser = (input: string) =>
    parseHelper<FormMLDeclaration>(services.FormMLDeclaration)(input)
      .then((x) => x.parseResult)
      .then((r) =>
        r.lexerErrors.length > 0 || r.parserErrors.length > 0
          ? Promise.reject(
              new Error(
                'Parsing failed with errors:\n' +
                  [...r.lexerErrors, ...r.parserErrors]
                    .map((e) => e.message)
                    .join('\n'),
              ),
            )
          : r.value,
      )

  describe('annotation', () => {
    test('declare annotation', async () => {
      const input = `annot fun newAnnotation()`
      const ast = await parser(input)
      expect(serialize(ast)).toMatchSnapshot()
    })

    test('declare annotation with one parameter', async () => {
      const input = `annot fun newAnnotation(name)`
      const ast = await parser(input)
      expect(serialize(ast)).toMatchSnapshot()
    })

    test('declare annotation with multiple parameters', async () => {
      const input = `annot fun newAnnotation(name, value)`
      const ast = await parser(input)
      expect(serialize(ast)).toMatchSnapshot()
    })

    test('parameters allow trailing comma', async () => {
      const input = `annot fun newAnnotation(name,)`
      const ast = await parser(input)
      expect(serialize(ast)).toMatchSnapshot()
    })

    test('parameters disallow more than one trailing commas', async () => {
      const input = `annot fun newAnnotation(name,,)`
      await expect(parser(input)).rejects.toThrow()
    })
  })
})
