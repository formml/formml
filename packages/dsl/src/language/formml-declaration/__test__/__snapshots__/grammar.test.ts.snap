// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formml declaration grammar > annotation > declare annotation 1`] = `
"{
  "$type": "FormMLDeclaration",
  "declarations": [
    {
      "$type": "AnnotationDeclaration",
      "name": "newAnnotation",
      "parameters": []
    }
  ]
}"
`;

exports[`formml declaration grammar > annotation > declare annotation with multiple parameters 1`] = `
"{
  "$type": "FormMLDeclaration",
  "declarations": [
    {
      "$type": "AnnotationDeclaration",
      "name": "newAnnotation",
      "parameters": [
        {
          "$type": "Parameter",
          "name": "name"
        },
        {
          "$type": "Parameter",
          "name": "value"
        }
      ]
    }
  ]
}"
`;

exports[`formml declaration grammar > annotation > declare annotation with one parameter 1`] = `
"{
  "$type": "FormMLDeclaration",
  "declarations": [
    {
      "$type": "AnnotationDeclaration",
      "name": "newAnnotation",
      "parameters": [
        {
          "$type": "Parameter",
          "name": "name"
        }
      ]
    }
  ]
}"
`;

exports[`formml declaration grammar > annotation > parameters allow trailing comma 1`] = `
"{
  "$type": "FormMLDeclaration",
  "declarations": [
    {
      "$type": "AnnotationDeclaration",
      "name": "newAnnotation",
      "parameters": [
        {
          "$type": "Parameter",
          "name": "name"
        }
      ]
    }
  ]
}"
`;

exports[`formml declaration grammar > annotation > type system > parameters can be primitive type "bool" 1`] = `
"{
  "$type": "FormMLDeclaration",
  "declarations": [
    {
      "$type": "AnnotationDeclaration",
      "name": "newAnnotation",
      "parameters": [
        {
          "$type": "Parameter",
          "name": "param1",
          "type": "bool"
        },
        {
          "$type": "Parameter",
          "name": "param2",
          "type": "bool"
        }
      ]
    }
  ]
}"
`;

exports[`formml declaration grammar > annotation > type system > parameters can be primitive type "datetime" 1`] = `
"{
  "$type": "FormMLDeclaration",
  "declarations": [
    {
      "$type": "AnnotationDeclaration",
      "name": "newAnnotation",
      "parameters": [
        {
          "$type": "Parameter",
          "name": "param1",
          "type": "datetime"
        },
        {
          "$type": "Parameter",
          "name": "param2",
          "type": "datetime"
        }
      ]
    }
  ]
}"
`;

exports[`formml declaration grammar > annotation > type system > parameters can be primitive type "decimal" 1`] = `
"{
  "$type": "FormMLDeclaration",
  "declarations": [
    {
      "$type": "AnnotationDeclaration",
      "name": "newAnnotation",
      "parameters": [
        {
          "$type": "Parameter",
          "name": "param1",
          "type": "decimal"
        },
        {
          "$type": "Parameter",
          "name": "param2",
          "type": "decimal"
        }
      ]
    }
  ]
}"
`;

exports[`formml declaration grammar > annotation > type system > parameters can be primitive type "num" 1`] = `
"{
  "$type": "FormMLDeclaration",
  "declarations": [
    {
      "$type": "AnnotationDeclaration",
      "name": "newAnnotation",
      "parameters": [
        {
          "$type": "Parameter",
          "name": "param1",
          "type": "num"
        },
        {
          "$type": "Parameter",
          "name": "param2",
          "type": "num"
        }
      ]
    }
  ]
}"
`;

exports[`formml declaration grammar > annotation > type system > parameters can be primitive type "text" 1`] = `
"{
  "$type": "FormMLDeclaration",
  "declarations": [
    {
      "$type": "AnnotationDeclaration",
      "name": "newAnnotation",
      "parameters": [
        {
          "$type": "Parameter",
          "name": "param1",
          "type": "text"
        },
        {
          "$type": "Parameter",
          "name": "param2",
          "type": "text"
        }
      ]
    }
  ]
}"
`;
